File: functions/src/lib/firestore/db/storeEvents.ts
import { firestore } from "../../../index"
import { serverTimestamp } from "../index"

export async function storeEvents(
  eventType: string,
  eventData: string,
  since_id?: number
) {
  const eventsRef = firestore.collection(`${eventType}Events`)

  const events = eventData
    .split("\n")
    .slice(1)
    .filter((e) => e !== "")
    .map((e) => JSON.parse(e.slice(6)))

  // - store in event log
  let batch = firestore.batch()
  for (const [i, event] of events.entries()) {
    if ((i + 1) % 500 === 0) {
      // - commit every 500 events
      await batch.commit()
      batch = firestore.batch()
    }
    const event_id = event.event_id

    if (since_id == event_id) continue
    const docRef = eventsRef.doc(event.event_id.toString())
    batch.set(docRef, {
      type: eventType,
      ...event,
      insertTimestamp: serverTimestamp(),
    })
  }

  await batch.commit()
}

File: functions/server/utils/store_events.py
from typing import List

from google.cloud import firestore

from .initialise_firestore import initialise_firestore

async def store_events(type: str, data: List[Event]):
    """Store events in the database.
    Args:
      type (str): The event type.
      data (str): The event data which is a stringified list of json.
    """
    # Get the database.
    db = initialise_firestore(use_async=True)

    event_data = json.loads(data)

    events_ref = db.collection("{eventType}Events")

    batch = db.batch()
    for index, event in enumerate(event_data):
        if index % 500 == 0:
            batch.commit()
            batch = db.batch()

        doc_ref = events_ref.document(str(event.event_id))

        batch.set(
            doc_ref,
            {
                **event_data,
                "type": type,
                "insertTimestamp": firestore.SERVER_TIMESTAMP,
            },
        )

    await batch.commit()
